### PR TITLE
Adding in a pipeline type param.

### PR DESCRIPTION
**Pull request by StewartW**
*Issue #, if available:* #185 

*Description of changes:*
I've re-introduced a pipeline type param, I hope you find this useful. The primary motivation is to make the overall shape of pipelines more configurable. For example we currently have pipelines at Skyscanner that are used for deploying CloudCustodian to all of our accounts, or executing CDK in them. (And I would like to eventually share some of these back to the ADF community at some point too.) 

I also think it is a good starting place to resolving issue #282 by creating a CDK pipeline type, and utilising the new CDK Pipeline Construct. 

By submitting this pull request, I confirm that you can use, modify, copy, and redistribute this contribution, under the terms of your choice.

### CODE REVIEW

File: src/lambda_codebase/initial_commit/bootstrap_repository/adf-build/shared/cdk/cdk_stacks/adf_standard_pipelines.py [hunk 1]
```diff
@@ -0,0 +1,260 @@
...
+
+    if stack_input["input"].get("params", {}).get("notification_endpoint"):
+        stack_input["input"]["topic_arn"] = adf_notifications.Notifications(
+            scope, "adf_notifications", stack_input["input"]
+        ).topic_arn
+
+    _source_name = generate_source_stage_for_pipeline(_stages, scope, stack_input)
+    generate_build_stage_for_pipeline(_stages, scope, stack_input)
+    print(stack_input["input"].get("environments", {}).get("targets", []))
+    for index, targets in enumerate(stack_input["input"].get("environments", {}).get("targets", [])):
```
Comment (sbkok) [inline src/lambda_codebase/initial_commit/bootstrap_repository/adf-build/shared/cdk/cdk_stacks/adf_standard_pipelines.py:39]: Would it be easy to move the target generation to its own function too?
In that case this function will become nothing else than setting up the notification endpoint, rendering source, build, and targets. Plus generating the related Pipeline at the end + webhooks for Github.

File: src/lambda_codebase/initial_commit/bootstrap_repository/adf-build/shared/cdk/cdk_stacks/adf_standard_pipelines.py [hunk 2]
```diff
@@ -0,0 +1,260 @@
...
+                stack_input["ssm_params"][ADF_DEPLOYMENT_REGION]["kms"],
+                stack_input["input"],
+                {},  # Empty target since this is a build only stage
+            ).build
+        )
+    elif "jenkins" in _build_name:
+        _stages.append(adf_jenkins.Jenkins(scope, "build", stack_input["input"]).build)
+
+
+def generate_source_stage_for_pipeline(_stages, scope, stack_input):
```
Comment (sbkok) [inline src/lambda_codebase/initial_commit/bootstrap_repository/adf-build/shared/cdk/cdk_stacks/adf_standard_pipelines.py:248]: Could you move the functions in the order as they will be executed, so source first, build second, targets third, actions fourth, etc?

File: src/lambda_codebase/initial_commit/bootstrap_repository/adf-build/shared/cdk/cdk_stacks/adf_standard_pipelines.py [hunk 3]
```diff
@@ -0,0 +1,260 @@
+# Copyright 2020 Amazon.com, Inc. or its affiliates. All Rights Reserved.
```
Comment (sbkok) [inline src/lambda_codebase/initial_commit/bootstrap_repository/adf-build/shared/cdk/cdk_stacks/adf_standard_pipelines.py:1]: On the filename, at the moment, it is called `adf_standard_pipelines.py` whereas you called the Pipelines Type "Default", maybe good to relate the two. Plus the name is plural: `pipelines`, whereas it is used to generate a single pipeline. What do you think about calling it `adf_default_pipeline.py` instead?

File: src/lambda_codebase/initial_commit/bootstrap_repository/adf-build/shared/cdk/cdk_stacks/tests/test_pipeline_creation.py [hunk 4]
```diff
@@ -0,0 +1,142 @@
+# Copyright 2020 Amazon.com, Inc. or its affiliates. All Rights Reserved.
```
Comment (sbkok) [inline src/lambda_codebase/initial_commit/bootstrap_repository/adf-build/shared/cdk/cdk_stacks/tests/test_pipeline_creation.py:1]: Thanks for adding tests for some of the changes you made, I appreciate this!

**Review by sbkok — COMMENTED** [review]
Thanks @StewartW for adding these changes and refactoring as you go.

File: src/lambda_codebase/initial_commit/bootstrap_repository/adf-build/shared/cdk/cdk_stacks/adf_standard_pipelines.py (hunk 3 above)
Comment (StewartW) [reply src/lambda_codebase/initial_commit/bootstrap_repository/adf-build/shared/cdk/cdk_stacks/adf_standard_pipelines.py:1]: Makes sense. 👍

**Comment by StewartW** [comment]
@sbkok  thanks for pulling in the latest patches. One less thing of my to-do list.

**Review by sbkok — APPROVED** [review]
Looks great @StewartW, thanks for adding this.

**Review by thomasmcgannon — APPROVED** [review]
LGTM

**Pull request merged**